Add needs-unwind to tests that depend on panicking

djkoloski opened the pull request:
These tests were found by running the test suite on fuchsia which compiles with `panic=abort` by default, then picking through the failures manually to locate the tests that require unwinding support.

Most of these tests are already opted-out on platforms that compile with `panic=abort` by default. This just generalizes it a bit more so that fuchsia tests can be run properly. Currently, the `needs-unwind` directive needs to be manually passed to compiletest (e.g. via `--test-args '--target-panic=abort'`). Eventually, I would like `x.py` or compiletest to determine whether the directive should be used automatically based on the target panic settings.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @Mark-Simulacrum (or someone else) soon.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

tmandry commented:
You should also be able to remove the existing `ignore-(platform)` lines that are there because the platform is built without unwinding support.

djkoloski commented:
I did that originally but the flag in compiletest needs a command-line argument (i.e. it's not inherently set for specific target configurations). Should we be pulling that setting from a config?

tmandry commented:
Ah right, the plan was to add a flag to bootstrap that enables `panic=abort` and tells compiletest about it. That should be a separate change. I think this is fine for now.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4372230528?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/ce8b18e077ae4656dba39a66cb59ee51f0de4b58/checks/4372230528/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

---- [ui] ui/closures/2229_closure_analysis/migrations/mir_calls_to_shims.rs stdout ----
diff of stderr:

1 error: changes to closure capture in Rust 2021 will affect which traits the closure implements
-   --> $DIR/mir_calls_to_shims.rs:20:38
3    |
4 LL |     let result = panic::catch_unwind(move || {
5    |                                      ^^^^^^^


11    |         --- in Rust 2018, this closure captures all of `f`, but in Rust 2021, it will only capture `f.0`
13 note: the lint level is defined here
-   --> $DIR/mir_calls_to_shims.rs:3:9
+   --> $DIR/mir_calls_to_shims.rs:4:9
15    |
---
3 #![deny(rust_2021_incompatible_closure_captures)]
4 //~^ NOTE: the lint level is defined here


The actual fixed differed from the expected fixed.
Actual fixed saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/closures/2229_closure_analysis/migrations/mir_calls_to_shims/mir_calls_to_shims.fixed
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
To only update this specific test, also pass `--test-args closures/2229_closure_analysis/migrations/mir_calls_to_shims.rs`

error: 2 errors occurred comparing output.
error: 2 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/closures/2229_closure_analysis/migrations/mir_calls_to_shims.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/closures/2229_closure_analysis/migrations/mir_calls_to_shims" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/closures/2229_closure_analysis/migrations/mir_calls_to_shims/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error: changes to closure capture in Rust 2021 will affect which traits the closure implements
   |
LL |     let result = panic::catch_unwind(move || {
   |                                      ^^^^^^^
   |                                      |
   |                                      |
   |                                      in Rust 2018, this closure implements `UnwindSafe` as `f` implements `UnwindSafe`, but in Rust 2021, this closure will no longer implement `UnwindSafe` because `f` is not fully captured and `f.0` does not implement `UnwindSafe`
   |                                      in Rust 2018, this closure implements `RefUnwindSafe` as `f` implements `RefUnwindSafe`, but in Rust 2021, this closure will no longer implement `RefUnwindSafe` because `f` is not fully captured and `f.0` does not implement `RefUnwindSafe`
LL |         f.0()
LL |         f.0()
   |         --- in Rust 2018, this closure captures all of `f`, but in Rust 2021, it will only capture `f.0`
note: the lint level is defined here
  --> /checkout/src/test/ui/closures/2229_closure_analysis/migrations/mir_calls_to_shims.rs:4:9
   |
LL | #![deny(rust_2021_incompatible_closure_captures)]
LL | #![deny(rust_2021_incompatible_closure_captures)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2021/disjoint-capture-in-closures.html>
help: add a dummy let to cause `f` to be fully captured
   |
LL ~     let result = panic::catch_unwind(move || {
LL +         let _ = &f;

error: aborting due to previous error


---
test result: FAILED. 12306 passed; 1 failed; 115 ignored; 0 measured; 0 filtered out; finished in 145.36s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "ui" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-12/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "12.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine extensions filecheck frontendopenacc frontendopenmp fuzzmutate globalisel hellonew hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:14:17
```

</details>

Mark-Simulacrum commented on src/test/ui/async-await/async-fn-size-moved-locals.rs:
```diff
@@ -8,6 +8,7 @@
 // See issue #59123 for a full explanation.
 
 // ignore-emscripten (sizes don't match)
+// ignore-fuchsia (sizes don't match exactly)
```
This seems unrelated to the PR's stated purpose, I'd prefer to split it (at least to a separate commit, ideally PR). I'm not sure why emscripten and fuchsia are special here, that seems surprising to me. (Maybe this is a needs-unwind in disguise?)

djkoloski replied:
This was accidentally included, I'll split this into a separate PR.

Mark-Simulacrum commented:
> I did that originally but the flag in compiletest needs a command-line argument (i.e. it's not inherently set for specific target configurations). Should we be pulling that setting from a config?

I forget what PR added needs-unwind (maybe my bad there), but if it's currently just not being used at all (which seems to be the case), it seems a little misleading to go through annotating tests with this flag, particularly when it "seems" like some of the ignore- flags would otherwise be removable.

I also noted that the current target specs for fuchsia don't have a panic strategy of abort, so it seems like in general unwinding is supported?

I am not strictly opposed to merging this, but I would like a better understanding of the expectation on your end for integrating needs-unwind into actual invocations of compile test from our build system (rough sense of timeline), and maybe some commentary in the PR description for how these tests were found (did you run with a panic=abort flag and just blanket added for where it failed?). My review comment suggests that it's possible some unrelated tests were also adjusted in this PR, which seems best to avoid and is part of the reason I'd like the test selection rationale better documented in the PR.

djkoloski commented:
> if it's currently just not being used at all (which seems to be the case), it seems a little misleading to go through annotating tests with this flag

There is a flag and it is functional, but it has to be manually passed (e.g. `./x.py test --test-args '--target-panic=abort'`). It should really be pulled from the target specification instead, but I haven't done any work to get that up and running yet. Once the flag is automatically specified, I would be comfortable removing the target-specific `ignore-` directives.

> I also noted that the current target specs for fuchsia don't have a panic strategy of abort, so it seems like in general unwinding is supported?

Although I believe it is now possible to build fuchsia with unwinding, I believe this was not always the case and the standard library is compiled with `panic=abort` by default. I can check with @tmandry on this.

> I would like a better understanding of the expectation on your end for integrating needs-unwind into actual invocations of compile test

I expect that eventually x.py will automatically determine whether the `needs-unwind` directive applies to a target. Getting the Rust test suite running on fuchsia is my primary objective.

> maybe some commentary in the PR description for how these tests were found

Updated.

> My review comment suggests that it's possible some unrelated tests were also adjusted in this PR

Ack, you are correct. Those are fixes for other tests that failed on fuchsia, but are unrelated to this PR. They were added by mistake and I've removed them from this PR.

djkoloski commented:
@rustbot ready

tmandry commented:
For context, my plan was to add an explicit flag to x.py for building with panic=abort (ideally with defaults for the platforms that don't support it). This would modify compilation flags (including `-Zpanic-abort-tests` in the case of testing) as well as cause `--target-panic=abort` to be passed to compiletest.

It's true that adding the flag while it's not fully functional could be misleading, but at least for now there wouldn't be any examples where it was there without the right platforms also being ignored explicitly. (Eventually we would take those out.)

Mark-Simulacrum commented:
I'm okay proceeding here -- it sounds like we're on a good path toward this being integrated, and it's already ~usable directly, when needed.

@djkoloski it looks like an unrelated commit and a couple merges landed on this branch -- can you rebase those away? After that's done I expect to r+

Mark-Simulacrum commented:
@bors r+

bors commented:
:pushpin: Commit ea68758299a89556e67f0bfffffb19c0c8346e8a has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum ea68758299a89556e67f0bfffffb19c0c8346e8a -->
<!-- homu: {"type":"Approved","sha":"ea68758299a89556e67f0bfffffb19c0c8346e8a","approver":"Mark-Simulacrum"} -->
